tensorflow/c/kernels_test.cc

```diff
@@ -161,6 +161,328 @@ TEST(TestKernel, TestRegisterKernelBuilder) {
+      .Attr("Attr: list(" #type ")")                          \
+      .SetShapeFn(tensorflow::shape_inference::UnknownShape)
+ATTR_TEST_REGISTER_OP(String, string);
+ATTR_TEST_REGISTER_OP(Int, int);
+ATTR_TEST_REGISTER_OP(Float, float);
+ATTR_TEST_REGISTER_OP(Bool, bool);
+ATTR_TEST_REGISTER_OP(Type, type);
+#undef ATTR_TEST_REGISTER_OP
+
+// Helper macros for the TF_OpKernelConstruction_GetAttr* tests.
+#define EXPECT_TF_META(attr_name, expected_list_size, expected_type,          \
+                       expected_total_size)                                   \
+  do {                                                                        \
+    auto m = TF_OpKernelConstruction_GetAttrMetadata(ctx, attr_name, status); \
+    EXPECT_EQ(TF_OK, TF_GetCode(status)) << TF_Message(status);               \
+    const unsigned char e = expected_list_size >= 0 ? 1 : 0;                  \
+    EXPECT_EQ(e, m.is_list);                                                  \
+    EXPECT_EQ(expected_list_size, m.list_size);                               \
+    EXPECT_EQ(expected_type, m.type);                                         \
+    EXPECT_EQ(expected_total_size, m.total_size);                             \
+  } while (0)
+
+typedef void* (*MyCreateFuncWithAttr)(TF_OpKernelConstruction*);
+class TestKernelAttr : public ::testing::Test {
+ public:
+  TestKernelAttr() {}
+  ~TestKernelAttr() {}
+
+  std::unique_ptr<OpKernel> GetFakeKernelWithAttr(const char* op_name,
+                                                  AttrValue v, Status* status) {
+    NodeDef def;
+    def.set_op(op_name);
+    def.set_name("FakeNode");
+    def.set_device("FakeDevice");
+    (*def.mutable_attr())["Attr"] = v;
+    return CreateOpKernel(DeviceType("FakeDevice"), nullptr, nullptr, def, 1,
+                          status);
+  }
+
+  void SetAttr(MyCreateFuncWithAttr MyCreateFuncAttr, const char* op_name,
+               AttrValue& v) {
+    TF_KernelBuilder* builder = TF_NewKernelBuilder(
+        op_name, "FakeDevice", MyCreateFuncAttr, &MyComputeFunc, &MyDeleteFunc);
+    {
+      TF_Status* status = TF_NewStatus();
+      TF_RegisterKernelBuilder("FakeNode", builder, status);
+      EXPECT_EQ(TF_OK, TF_GetCode(status));
+      TF_DeleteStatus(status);
+    }
+    Status status;
+    std::unique_ptr<OpKernel> kernel =
+        GetFakeKernelWithAttr(op_name, v, &status);
+    TF_EXPECT_OK(status);
+    ASSERT_NE(nullptr, kernel.get());
+    kernel->Compute(nullptr);
+
+    ASSERT_TRUE(delete_called);
+  }
+};
+
+TEST_F(TestKernelAttr, String) {
+  auto my_create_func = [](TF_OpKernelConstruction* ctx) {
+    struct MyCustomKernel* s = new struct MyCustomKernel;
+    s->created = true;
+    s->compute_called = false;
+
+    std::unique_ptr<char[]> val(new char[5]);
+    TF_Status* status = TF_NewStatus();
+    EXPECT_TF_META("Attr", -1, TF_ATTR_STRING, 5);
+    TF_OpKernelConstruction_GetAttrString(ctx, "Attr", val.get(), 5, status);
```

Comment:
Can you label non-obvious input parameters to `EXPECT_TF_META` and `TF_OpKernelConstruction_GetAttrString`. I see that we don't label them in many cases in similar tests in `c_api_test.cc` but it would be much easier to read with labels.

Example:
```cpp
EXPECT_TF_META(/*attr_name*/"Attr", /*expected_list_size*/-1, /*expected_type*/TF_ATTR_STRING, /*expected_total_size*/5);
TF_OpKernelConstruction_GetAttrString(ctx, "Attr", val.get(), /*max_length*/5, status);
```

Reply:
done